### PR TITLE
Restore chunked loss FSDP flags on errors

### PR DESCRIPTION
**Pull request by fallintoplace**
## Summary

Make the chunked CE loss FSDP flag changes exception-safe.

## Root cause

`ChunkedCELoss` temporarily disables lm_head FSDP resharding and gradient sync while it processes sequence chunks. The old code restored those flags only after the loop completed. If `lm_head(h_chunk)`, loss computation, or `chunk_loss.backward()` raised, the restore block was skipped, leaving the lm_head in a mutated FSDP runtime state.

## Changes

- Add a small context manager around the temporary lm_head FSDP state changes.
- Restore the exact prior FSDP param-group state in `finally`, rather than always restoring to `True`.
- Keep the existing final `reshard()` cleanup behavior.
- Add a CPU unit test with a fake FSDP module that raises inside the chunk loop and verifies state restoration.

## Validation

- `python -m pytest -q tests/unit_tests/test_loss.py::TestChunkedCELoss::test_fsdp_flags_are_restored_if_chunk_loop_raises`
- `python -m pytest -q tests/unit_tests/test_loss.py`
- `pre-commit run --files torchtitan/components/loss.py tests/unit_tests/test_loss.py`

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -444,6 +445,66 @@ def result(self) -> torch.Tensor:
         return self._buffer
 
 
+@contextmanager
+def _temporary_lm_head_fsdp_no_sync(lm_head):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:449]: Is this lm_head specific or chunked loss specific?

If not, maybe we should put it in torchtitan/distributed/fsdp.py or pytorch core

Also, the name should be something like `disable_fsdp_reshard`

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -571,44 +632,35 @@ def _chunk(t):
...
+                if fsdp_enabled
+                else nullcontext()
+            )
+            with fsdp_context:
+                last_idx = len(h_chunks) - 1
+                for i, (h_chunk, label_chunk) in enumerate(zip(h_chunks, label_chunks)):
+                    if fsdp_enabled and i == last_idx:
+                        lm_head.set_requires_gradient_sync(  # pyrefly: ignore[not-callable]
+                            True, recurse=False
+                        )
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:646]: seems still need to keep special logic regarding fsdp because of chunked loss specific details

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -444,6 +445,66 @@ def result(self) -> torch.Tensor:
...
+        (fsdp_param_group.reduce_grads, fsdp_param_group.all_reduce_grads)
+        for fsdp_param_group in lm_head_state._fsdp_param_groups
+    ]
+
+    try:
+        lm_head.set_reshard_after_forward(False)
+        lm_head.set_reshard_after_backward(False)
+        lm_head.set_requires_gradient_sync(False, recurse=False)
+        yield
+    finally:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:489]: > If lm_head(h_chunk), loss computation, or chunk_loss.backward() raised, the restore block was skipped, leaving the lm_head in a mutated FSDP runtime state.

If the thing you mentioned raises
1. what's the motivation to restore FSDP runtime?
2. are we sure that the variables you are accessing in `finally` is still legit?

**Review by tianyu-l — CHANGES_REQUESTED** [review]
@weifengpy could you take a look?

File: torchtitan/components/loss.py (hunk 1 above)
Comment (weifengpy) [reply torchtitan/components/loss.py:449]: torch fully_shard might be a better place to host this, say `FSDPModule.defer_reshard_and_grad_sync()`, just like `FSDPModule.reset_iter_state()`

in titan, it should be
```
  with lm_head.defer_reshard_and_grad_sync() as scope:
      for i, chunk in enumerate(chunks):
          if i == last_idx:
              scope.set_requires_gradient_sync(True)
          ...
```

in torch fully_shard, it's like this
``` 
  @contextmanager
  def defer_reshard_and_grad_sync(self):
      saved = _capture_fsdp_runtime_policy(self, recurse=True)
      scope = _DeferredReshardAndGradSyncScope(self)

      try:
          self.set_reshard_after_forward(False, recurse=True)
          self.set_reshard_after_backward(False, recurse=True)
          self.set_requires_gradient_sync(False, recurse=False)
          yield scope
      finally:
          _restore_fsdp_runtime_policy(saved)
          self.reshard()
```

what's the motivation for recovery? do you want to continue the training when error happens?

**Review by weifengpy — COMMENTED** [review]
left comment about motivation
